c++/src/kj/refcount.c++

```diff
@@ -35,4 +38,34 @@ void Refcounted::disposeImpl(void* pointer) const {
   }
 }
 
+// =======================================================================================
+// Atomic (thread-safe) refcounting
+
+AtomicRefcounted::~AtomicRefcounted() noexcept(false) {
+  KJ_ASSERT(refcount == 0, "Refcounted object deleted with non-zero refcount.");
+}
+
+void AtomicRefcounted::disposeImpl(void* pointer) const {
+  if (__atomic_sub_fetch(&refcount, 1, __ATOMIC_RELEASE) == 0) {
+    __atomic_thread_fence(__ATOMIC_ACQUIRE);
```

Comment:
Ah, hmm, I guess using `ACQ_REL` would burden both if-paths with the acquire fence, whereas as written it only has an acquire fence if the refcount hits 0.

Reply:
Right. (I think I copied this pattern from some refcounting implementation that knows what it's going, in any case...)

Reply:
(The reason I can't remember if I copied it is because this is actually old code that used to live in the Workers implementation until Kyle convinced me to use shared_ptr... so I actually wrote it like six months ago.)